### PR TITLE
Windows daemon manager

### PR DESCRIPTION
**Pull request by saurabhc123**
### Summary
This PR adds the `DaemonManager` and `ManagedDaemon` components to support EBS-TaskAttach for Windows.

### Implementation details
We refactored the existing `DaemonManager` and `ManagedDaemon` modules for linux and repurposed the high-level workflow. The platform specific details were moved to the respective windows/linux modules.

### Testing
We used the existing linux tests and refactored them so that all of them can be repurposed for Windows. 

New tests cover the changes: `Yes`

### Description for the changelog
Add support EBS-TaskAttach for Windows.

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/ebs/watcher_windows.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package ebs
+
+const (
+	hostMountDir = "C:\\var\\lib\\kubelet"
```
Comment (saurabhc123) [inline agent/ebs/watcher_windows.go:20]: @jiuchoe4 - Can you fix this path as part of your testing?

File: agent/api/task/task.go [hunk 2]
```diff
@@ -1747,7 +1747,6 @@ func (task *Task) dockerConfig(container *apicontainer.Container, apiVersion doc
 		Env:          dockerEnv,
 	}
 
-	// TODO [SC] - Move this as well as 'dockerExposedPorts' SC-specific logic into a separate file
```
Comment (mye956) [inline agent/api/task/task.go:1750]: Q: Just curious, any reason why we're removing this TODO comment here? Not sure if I have the full context for this.

File: agent/ebs/watcher_linux.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package ebs
+
+const (
+	csiDriverSocketAddress = "/var/run/ecs/ebs-csi-driver/csi-driver.sock"
```
Comment (mye956) [inline agent/ebs/watcher_linux.go:20]: Is this being used anywhere? I know we use `hostMountDir` that's already being used in the existing implementation and we're declaring the value of the variable differently between linux and windows but I don't recall if we use `csiDriverSocketAddress`.

File: agent/ebs/watcher_linux_test.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package ebs
+
+const (
+	CSIDriverSocketPath = "/var/run/ecs/ebs-csi-driver/csi-driver.sock"
```
Comment (mye956) [inline agent/ebs/watcher_linux_test.go:20]: nit: Any reason why we need to make this variable global? Seems like we're just using this within the same package.

File: agent/ebs/watcher_windows.go (hunk 1 above)
Comment (mye956) [inline agent/ebs/watcher_windows.go:20]: Q: Do we also need a `csiDriverSocketAddress` here for windows?

File: agent/engine/daemonmanager/daemon_manager.go [hunk 5]
```diff
@@ -44,3 +61,168 @@ func NewDaemonManager(manageddaemon *md.ManagedDaemon) DaemonManager {
...
+			// https://github.com/moby/moby/blob/master/api/types/mount/mount.go#L52
+			bindOptions = dockermount.BindOptions{Propagation: dockermount.PropagationShared}
+		}
+		logger.Info(fmt.Sprintf("bindMount Propagation: %s", bindOptions.Propagation),
+			logger.Fields{
+				field.Image: loadedImageRef,
+			})
+		typeBind := dockermount.TypeBind
+		mountPoint := dockermount.Mount{}
+		if mp.SourceVolumeType == "npipe" {
```
Comment (mye956) [inline agent/engine/daemonmanager/daemon_manager.go:118]: Q: This is essentially only applicable for windows, is that correct?

File: agent/engine/daemonmanager/daemon_manager_linux.go [hunk 6]
```diff
@@ -17,191 +17,14 @@
...
 
 const (
-	daemonUID                                   = 0
-	daemonMountPermission           fs.FileMode = 0755
-	daemonLogPermission             fs.FileMode = 0777
-	ecsAgentLogFileENV                          = "ECS_LOGFILE"
-	defaultECSAgentLogPathContainer             = "/log"
-	socketPathHostRoot                          = "/var/run/ecs"
-	logPathHostRoot                             = "/log/daemons"
+	daemonUID                                     = 0
```
Comment (mye956) [inline agent/engine/daemonmanager/daemon_manager_linux.go:24]: nit (non-blocking): I'm seeing a lot of different config files that's scattered in different files. Perhaps we can try to scope it down a bit to having fewer config files.

File: agent/ebs/watcher_windows.go (hunk 1 above)
Comment (jiuchoe4) [reply agent/ebs/watcher_windows.go]: Changed and merged this as a part of commit [82fc629](https://github.com/aws/amazon-ecs-agent/commit/82fc62937a73c3e30ea3549112e6b592807955e0)

**Review by mye956 — DISMISSED** [review]

File: agent/api/task/task.go (hunk 2 above)
Comment (saurabhc123) [reply agent/api/task/task.go:1750]: This was unintentional. I will add that back in.

File: agent/ebs/watcher_linux.go (hunk 3 above)
Comment (saurabhc123) [reply agent/ebs/watcher_linux.go:20]: You are correct. I will remove this file.

File: agent/ebs/watcher_linux_test.go (hunk 4 above)
Comment (saurabhc123) [reply agent/ebs/watcher_linux_test.go]: This constant and the file will be removed. It's not needed anymore.

File: agent/ebs/watcher_windows.go (hunk 1 above)
Comment (saurabhc123) [reply agent/ebs/watcher_windows.go]: We load the socket path from the default Windows config in `\agent\config\config_windows.go`

File: agent/engine/daemonmanager/daemon_manager_linux.go (hunk 6 above)
Comment (saurabhc123) [reply agent/engine/daemonmanager/daemon_manager_linux.go:24]: I have refactored as much as possible. For some constants that are specific to a module, I have kept it to the module itself. Most are limited to the module. For the Linux ones, I refactored as much as possible, but didn't want to perturb the code too much. Let me know if you see any config that can be made global.

**Review by mye956 — APPROVED** [review]

**Review by xxx0624 — COMMENTED** [review]
I mainly took a look at the ecs-agent changes which look good to me.

**Review by xxx0624 — APPROVED** [review]

**Pull request merged**